File: scilab/modules/gui/sci_gateway/c/sci_xgetfile.c
/*
 * Scilab ( http://www.scilab.org/ ) - This file is part of Scilab
 * Copyright (C) 2006 - INRIA - Allan CORNET
 * Copyright (C) 2007 - INRIA - Vincent COUVERT (java version)
 * 
 * This file must be used under the terms of the CeCILL.
 * This source file is licensed as described in the file COPYING, which
 * you should have received as part of this distribution.  The terms
 * are also available at    
 * http://www.cecill.info/licences/Licence_CeCILL_V2-en.txt
 *
 */
 
#include "gw_gui.h"
#include "machine.h"
#include "stack-c.h"
#include "CallFileChooser.h"
#include "MALLOC.h"
#include "localization.h"
#include "Scierror.h"
#include "cluni0.h"
/*--------------------------------------------------------------------------*/
int C2F(sci_xgetfile)(char *fname,unsigned long fname_len)
{
  int fileChooserID = 0;
  
  int nbRow = 0, nbCol = 0;

  char **userSelection = NULL;
  int voidSelectionAdr = 0;

  char *optName = NULL;

  int titleAdr = 0;
  int initialDirectoryAdr = 0;
  int fileMaskAdr = 0;

  char *title = NULL, *fileMask = NULL, *initialDirectory = NULL;

  CheckRhs(0,3);
  CheckLhs(1,1);

  if ((optName = (char*)MALLOC(sizeof(char*)*(strlen("title")+1))) == NULL)
    {
      Scierror(999, _("%s: No more memory.\n"),fname);
      return FALSE;
    }
  
  
  if (Rhs==1)
    {
      /* Argument is file_mask or optional title */
      if (VarType(1) == sci_strings) /* Check type */
        {
          if (IsOpt(1, optName)) /* Check if optional title */
            {
              if (strcmp(optName,"title")==0) /* Check if optional parameter is title */
                {
                  GetRhsVar(1, STRING_DATATYPE, &nbRow, &nbCol, &titleAdr);
                  if (nbCol !=1)
                    {
                      Scierror(999, _("%s: Wrong size for input argument #%d: A string expected.\n"),fname, 1);
                      FREE(optName);
                      return FALSE;
                    }
                  title = cstk(titleAdr);
                }
              else
                {
                  Scierror(999, _("%s: Wrong name for optional input argument #%d: '%s' expected.\n"),fname, 1, "title");
                  FREE(optName);
                  return FALSE;
                }
            }
          else
            {
              GetRhsVar(1, STRING_DATATYPE, &nbRow, &nbCol, &fileMaskAdr);
              if (nbCol !=1)
                {
                  Scierror(999, _("%s: Wrong size for input argument #%d: A string expected.\n"),fname, 1);
                  FREE(optName);
                  return FALSE;
                }
              fileMask = cstk(fileMaskAdr);
            }
        }
      else
        {
          Scierror(999, _("%s: Wrong type for input argument #%d: A string expected.\n"),fname, 1);
          FREE(optName);
          return FALSE;
        }
    }
  else if (Rhs==2)
    {
      /* First argument is file_mask */
      if (VarType(1) == sci_strings) /* Check type */
        {
          GetRhsVar(1, STRING_DATATYPE, &nbRow, &nbCol, &fileMaskAdr);
          if (nbCol !=1)
            {
              Scierror(999,  _("%s: Wrong size for input argument #%d: A string expected.\n"),fname,1);
              FREE(optName);
              return FALSE;
            }
          fileMask = cstk(fileMaskAdr);
        }
      else
        {
          Scierror(999,  _("%s: Wrong type for input argument #%d: A string expected.\n"),fname, 1);
          FREE(optName);
          return FALSE;
        }

      /* Second argument is initial directory or optional title */
      if (VarType(2) == sci_strings) /* Check type */
        {
          if (IsOpt(2, optName)) /* Check if optional title */
            {
              if (strcmp(optName,"title")==0) /* Check if optional parameter is title */
                {
                  GetRhsVar(2, STRING_DATATYPE, &nbRow, &nbCol, &titleAdr);
                  if (nbCol !=1)
                    {
                      Scierror(999,  _("%s: Wrong size for input argument #%d: A string expected.\n"),fname, 2);
                      FREE(optName);
                      return FALSE;
                    }
                  title = cstk(titleAdr);
                }
              else
                {
                  Scierror(999, _("%s: Wrong name for optional input argument #%d: '%s' expected.\n"),fname, 2, "title");
                  FREE(optName);
                  return FALSE;
                }
            }
          else
            {
              GetRhsVar(2, STRING_DATATYPE, &nbRow, &nbCol, &initialDirectoryAdr);
              if (nbCol !=1)
                {
                  Scierror(999,  _("%s: Wrong size for input argument #%d: A string expected.\n"),fname, 2);
                  FREE(optName);
                  return FALSE;
                }
              initialDirectory = cstk(initialDirectoryAdr);
            }
        }
      else
        {
          Scierror(999, _("%s: Wrong type for input argument #%d: A string expected.\n"),fname, 2);
          FREE(optName);
          return FALSE;
        }
    }
  else if (Rhs==3)
    {
      /* First argument is file_mask */
      if (VarType(1) == sci_strings) /* Check type */
        {
          GetRhsVar(1, STRING_DATATYPE, &nbRow, &nbCol, &fileMaskAdr);
          if (nbCol !=1)
            {
              Scierror(999, _("%s: Wrong size for input argument #%d: A string expected.\n"),fname, 1);
              FREE(optName);
              return FALSE;
            }
          fileMask = cstk(fileMaskAdr);
        }
      else
        {
          Scierror(999,  _("%s: Wrong type for input argument #%d: A string expected.\n"),fname, 1);
          return FALSE;
        }

      /* Second argument is initial directory */
      if (VarType(2) == sci_strings) /* Check type */
        {
          GetRhsVar(2, STRING_DATATYPE, &nbRow, &nbCol, &initialDirectoryAdr);
          if (nbCol !=1)
            {
              Scierror(999, _("%s: Wrong size for input argument #%d: A string expected.\n"),fname, 2);
              FREE(optName);
              return FALSE;
            }
          initialDirectory = cstk(initialDirectoryAdr);
        }
      else
        {
          Scierror(999, _("%s: Wrong type for input argument #%d: A string expected.\n"),fname, 2);
          FREE(optName);
          return FALSE;
        }
     
      /* Third argument is title or optional title */
      if (VarType(3) == sci_strings)  /* Check type */
        {
          if (IsOpt(3, optName)) /* Check if optional title */
            {
              if (strcmp(optName,"title")==0) /* Check if optional parameter is title */
                {
                  GetRhsVar(3, STRING_DATATYPE, &nbRow, &nbCol, &titleAdr);
                  if (nbCol !=1)
                    {
                      Scierror(999, _("%s: Wrong type for input argument #%d: A string expected.\n"),fname, 3);
                      FREE(optName);
                      return FALSE;
                    }
                  title = cstk(titleAdr);
                }
              else
                {
                  Scierror(999, _("%s: Wrong name for optional input argument #%d: '%s' expected.\n"),fname, 3, "title");
                  FREE(optName);
                  return FALSE;
                }
            }
          else
            {
              GetRhsVar(3, STRING_DATATYPE, &nbRow, &nbCol, &titleAdr);
              if (nbCol !=1)
                {
                  Scierror(999, _("%s: Wrong size for input argument #%d: A string expected.\n"),fname, 3);
                  FREE(optName);
                  return FALSE;
                }
              title = cstk(titleAdr);
            }
        }
      else
        {
          Scierror(999, _("%s: Wrong type for input argument #%d: A string expected.\n"),fname, 3);
          FREE(optName);
          return FALSE;
        }
    }

  /* Create the Java Object */
  fileChooserID = createFileChooser();

  /* Do not allow to select a directory */
  setFileChooserFileSelectionOnly(fileChooserID);
  
  /* Set options */
  if (title != NULL)
    {
      setFileChooserTitle(fileChooserID, title);
    }

  if (initialDirectory != NULL)
    {
		char expandedpath[PATH_MAX+1];
		int out_n = 0;
		C2F(cluni0)(initialDirectory ,expandedpath, &out_n,(int)strlen(initialDirectory),PATH_MAX);
		setFileChooserInitialDirectory(fileChooserID, expandedpath);
    }

  if (fileMask != NULL)
    {
      setFileChooserMask(fileChooserID, fileMask);
    }

  /* Display it and wait for a user input */
  fileChooserDisplayAndWait(fileChooserID);

  /* Read the size of the selection, if 0 then no file selected */
  nbRow = getFileChooserSelectionSize(fileChooserID);

  if (nbRow !=0 ) 
    {
      /* The user selected a file --> returns the files names */
      nbCol = 1;
      
      userSelection = getFileChooserSelection(fileChooserID);

      CreateVarFromPtr(Rhs+1, MATRIX_OF_STRING_DATATYPE, &nbRow, &nbCol, userSelection);
    }
  else
    {
      /* The user canceled the selection --> returns an empty matrix */
      nbRow = 0;
      nbCol = 0;

      CreateVar(Rhs+1, STRING_DATATYPE, &nbRow, &nbCol, &voidSelectionAdr);
    }
  
  LhsVar(1)=Rhs+1;
  C2F(putlhsvar)();

  FREE(optName);

  return TRUE;
}
/*--------------------------------------------------------------------------*/

File: scilab/modules/gui/sci_gateway/fortran/sci_xgetfile.f
c Scilab ( http://www.scilab.org/ ) - This file is part of Scilab
c Copyright (C) INRIA
c 
c This file must be used under the terms of the CeCILL.
c This source file is licensed as described in the file COPYING, which
c you should have received as part of this distribution.  The terms
c are also available at    
c http://www.cecill.info/licences/Licence_CeCILL_V2-en.txt

      subroutine intsxgetfile(fname)
C     ================================================================
c     SCILAB function : xgetfile, fin = 1
C     ================================================================
       character*(*) fname
       integer topk,rhsk,topl
       logical checkrhs,checklhs,cresmat2,getsmat,checkval,bufstore
       logical crepointer,crestring,isoptlw,getwsmat
       include 'stack.h'
       character name*(nlgh)
c
       integer iadr, sadr
       iadr(l)=l+l-1
       sadr(l)=(l/2)+1
       rhs = max(0,rhs)
c
       lbuf = 1
       topk = top 
       rhsk = rhs 
       if(.not.checkrhs(fname,0,3)) return
       if(.not.checklhs(fname,1,1)) return
c       checking variable a (number 1)
c       
       if(rhs .le. 0) then
        top = top+1
        rhs = rhs+1
        nlr1 = 3
        if(.not.cresmat2(fname,top,nlr1,lr1)) return
        call cvstr(nlr1,istk(lr1),'*.*',0)
       endif
c       checking variable dirname (number 2)
c       
       if(rhs .le. 1) then
        top = top+1
        rhs = rhs+1
        nlr2 = 1
        if(.not.cresmat2(fname,top,nlr2,lr2)) return
        call cvstr(nlr2,istk(lr2),'.',0)
       endif

       if(rhs .le. 2) then
        top = top+1
        rhs = rhs+1
        nlr3 = 9
        if(.not.cresmat2(fname,top,nlr3,lr3)) return
        call cvstr(nlr3,istk(lr3),'File menu',0)
       endif

       if(.not.crepointer(fname,top+1,lw3)) return


       if (rhsk.eq.3) then 
          idir=1
          if(.not.getsmat(fname,top,top-rhs+1,m1,n1,1,1,lr1,nlr1))
     $         return
          if(.not.checkval(fname,m1*n1,1)) return
          if(.not.bufstore(fname,lbuf,lbufi1,lbuff1,lr1,nlr1)) return

          if(.not.getsmat(fname,top,top-rhs+2,m2,n2,1,1,lr2,nlr2))
     $         return
          if(.not.checkval(fname,m2*n2,1)) return
          if(.not.bufstore(fname,lbuf,lbufi2,lbuff2,lr2,nlr2)) return

          if (.not.getwsmat(fname,top,top-rhs+3,m3,n3,il3,ild3)) return

          call xgetfile(buf(lbufi1:lbuff1),buf(lbufi2:lbuff2),
     $         stk(lw3),ne5,err,idir,istk(il3),istk(ild3),m3*n3)
       endif
       if (rhsk.eq.2) then 
          if(.not.getsmat(fname,top,top-rhs+1,m1,n1,1,1,lr1,nlr1))
     $         return
          if(.not.checkval(fname,m1*n1,1)) return
          if(.not.bufstore(fname,lbuf,lbufi1,lbuff1,lr1,nlr1)) return

          if (isoptlw(top,top-rhs+2,name)) then 
             if (name.ne.'title') then 
                buf = fname // 'optional argument must be title='
                call error(999)
                return
             endif
             idir = 0
             if (.not.getwsmat(fname,top,top-rhs+2,m2,n2,il2,ild2))
     $            return
             call xgetfile(buf(lbufi1:lbuff1),'.',
     $            stk(lw3),ne5,err,idir,istk(il2),istk(ild2),m2*n2)
          else
             idir = 1
             if(.not.getsmat(fname,top,top-rhs+2,m2,n2,1,1,lr2,nlr2))
     $            return
             if(.not.checkval(fname,m2*n2,1)) return
             if(.not.bufstore(fname,lbuf,lbufi2,lbuff2,lr2,nlr2)) return

             if (.not.getwsmat(fname,top,top-rhs+3,m3,n3,il3,ild3))
     $            return
             call xgetfile(buf(lbufi1:lbuff1),buf(lbufi2:lbuff2),
     $            stk(lw3),ne5,err,idir,istk(il3),istk(ild3),m3*n3)
          endif
       endif
       if ( rhsk.eq.1) then 
          idir = 0
          if (isoptlw(top,top-rhs+1,name)) then 
             if (name.ne.'title') then 
                buf = fname // ' optional argument must be title='
                call error(999)
                return
             endif
             if (.not.getwsmat(fname,top,top-rhs+1,m1,n1,il1,ild1))
     $            return
             call xgetfile('*.*','.',
     $            stk(lw3),ne5,err,idir,istk(il1),istk(ild1),m1*n1)
          else
             if(.not.getsmat(fname,top,top-rhs+1,m1,n1,1,1,lr1,nlr1))
     $            return
             if(.not.checkval(fname,m1*n1,1)) return
             if(.not.bufstore(fname,lbuf,lbufi1,lbuff1,lr1,nlr1)) return

             if(.not.getsmat(fname,top,top-rhs+2,m2,n2,1,1,lr2,nlr2))
     $            return
             if(.not.checkval(fname,m2*n2,1)) return
             if(.not.bufstore(fname,lbuf,lbufi2,lbuff2,lr2,nlr2)) return

             if (.not.getwsmat(fname,top,top-rhs+3,m3,n3,il3,ild3))
     $            return
             call xgetfile(buf(lbufi1:lbuff1),buf(lbufi2:lbuff2),
     $            stk(lw3),ne5,err,idir,istk(il3),istk(ild3),m3*n3)
          endif
       endif
       if (rhsk.le.0 ) then 
          idir=0
          if(.not.getsmat(fname,top,top-rhs+1,m1,n1,1,1,lr1,nlr1))
     $         return
          if(.not.checkval(fname,m1*n1,1)) return
          if(.not.bufstore(fname,lbuf,lbufi1,lbuff1,lr1,nlr1)) return
          
          if(.not.getsmat(fname,top,top-rhs+2,m2,n2,1,1,lr2,nlr2))
     $         return
          if(.not.checkval(fname,m2*n2,1)) return
          if(.not.bufstore(fname,lbuf,lbufi2,lbuff2,lr2,nlr2)) return
          
          if (.not.getwsmat(fname,top,top-rhs+3,m3,n3,il3,ild3))
     $         return
          call xgetfile(buf(lbufi1:lbuff1),buf(lbufi2:lbuff2),
     $         stk(lw3),ne5,err,idir,istk(il3),istk(ild3),m3*n3)
       endif
C
C      the result is a string and xgetfile 
C      will have to create ( malloc ) space for storing the result
C      we will keep track of this allocated string through lw3.

       if(err .gt. 0) then 
        buf = fname // 'Internal Error' 
        call error(999)
       endif
c
       topk=top-rhs
       topl=top+1
c     
       if(lhs .ge. 1) then
c       --------------output variable: res
        top=topl+1
        if(.not.crestring(fname,top,ne5,ilrs)) return
        call ccharf(ne5,stk(lw3),istk(ilrs))
       endif
c     Putting in order the stack
       if(lhs .ge. 1) then
        call copyobj(fname,topl+1,topk+1)
       endif
       top=topk+lhs
       return
       end
C     ================================================================
